---
title: "MICB475_R_code"
authors: "IP,AW,TW,SD,SK"
date: "2023-11-07"
output: html_document
function: performs analyses on microbiome data for alpha+beta diversity, differential expression (taxonomic+functional). Code is broken into sections according to function, and is set to store all figures as .png files in the accompanying "figures" directory. It can be run to reproduce all figures and statistics contained within the submitted manuscript.
organisation: Alpha diversity (line 54), Beta diversity (line 170), DESeq2 analysis (line 248), and PICRUSt2 analysis (line 345)
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Setting up =  import packages and data
```{r,echo=FALSE,message=FALSE}
#!/bin/R
### Install all packages
library(apeglm)
library(DESeq2)
library(ggh4x)
library(ggpicrust2)
library(ggprism)
library(patchwork)
library(picante)
library(readr)
library(tibble)
library(tidyverse)

#read abundance data
abundance_file <- "picrust2_out_pipeline/pathways_out/path_abun_unstrat.tsv"
abundance_data <- read_delim(abundance_file, delim = "\t", col_names = TRUE, trim_ws = TRUE) %>% as.data.frame()
#read metadata
metadata <- read_delim("colombia/metadata_categorized_CL.txt",delim = "\t",escape_double = FALSE,trim_ws = TRUE,show_col_types = FALSE)

#isolate smoking and nonsmoking from abundance data and metadata
metadata_smoking <- filter(metadata, metadata$smoker=="Yes")
metadata_nonsmoking <- filter(metadata, metadata$smoker=="No")

abundance_data_smoking <- select(abundance_data, metadata_smoking$`#SampleID`)
abundance_data_smoking$pathway = abundance_data$pathway
abundance_data_smoking <- abundance_data_smoking %>% select(pathway, everything())

abundance_data_nonsmoking <- select(abundance_data, metadata_nonsmoking$`#SampleID`)
abundance_data_nonsmoking$pathway = abundance_data$pathway
abundance_data_nonsmoking <- abundance_data_nonsmoking %>% select(pathway, everything())

### Load phyloseq object and subset according to smoker status
load("phyloseq_object_final.RData") 
phylo_smoking <-  subset_samples(phyloseq_object_final, smoker == "Yes")
phylo_nonsmoking <-  subset_samples(phyloseq_object_final, smoker == "No")
taxa_info = as.data.frame(tax_table(phylo_smoking))
taxa_info$ASV = rownames(taxa_info)
```

# ALPHA DIVERSITY ANALYSIS  ##############################
### Alpha diversity plots and stats for LDL (Figure 1)
```{r, message=FALSE}
load("phyloseq_object_final_highcount.RData")
estimate_richness(phyloseq_object_final_highcount)
phylo_dist <- pd(t(otu_table(phyloseq_object_final_highcount)), phy_tree(phyloseq_object_final_highcount),
                 include.root=F)
sample_data(phyloseq_object_final_highcount)$PD <- phylo_dist$PD

### Categorise data by smoking and diet for analysis
sample_data(phyloseq_object_final_highcount)[["legend2"]] <- paste(sample_data(phyloseq_object_final_highcount)[[32]], sample_data(phyloseq_object_final_highcount)[[38]])
sample_data(phyloseq_object_final_highcount)[["legend"]] = paste(sample_data(phyloseq_object_final_highcount)[[32]],sample_data(phyloseq_object_final_highcount)[[37]])

### For the manuscript - total count within each category
paste("n('Smoking') = ",length(meta$smoker[meta$smoker == 'Yes']))
paste("n('Nonsmoking') = ",length(meta$smoker[meta$smoker == 'No']))
paste("n(Smoking, high fiber) = ",length(meta$legend2[meta$legend2 == 'Yes high']))
paste("n(Smoking, low fiber) = ",length(meta$legend2[meta$legend2 == 'Yes low']))
paste("n('Nonsmoking high fiber') = ",length(meta$legend2[meta$legend2 == 'No high']))
paste("n('Nonsmoking low fiber') = ",length(meta$legend2[meta$legend2 == 'No low']))
paste("n('Smoking high LDL') = ",length(meta$legend[meta$legend == 'Yes high']))
paste("n('Smoking low LDL') = ",length(meta$legend[meta$legend == 'Yes low']))

### Define colours and diversity metrics for graph
sample_colors <- c("lightblue", "maroon", "olivedrab","orange")
diversity_measures <- c("Observed", "Shannon", "Chao1", "ACE", "Simpson", "InvSimpson", "Fisher")

### Generate alpha-diversity plot for LDL categories 
alpha_LDL <- phyloseq_object_final_highcount %>%                                                             
  plot_richness(x = "legend",measures = diversity_measures) +                             
  geom_boxplot(aes(fill = legend), show.legend = FALSE)+             #make violin plot, set fill aes to sampletype
  theme_linedraw()+                                                     #change theme to classic
  xlab(NULL)+                                                           #no label on x-axis
  theme(axis.text.y.left = element_text(size = 20),                     #adjust y-axis text
        axis.text.x = element_text(size = 20, hjust = 0.5, vjust = 1, angle = -90),           #adjust x-axis label position
        axis.title.y = element_text(size = 20))+                        #adjust y-axis title
  theme(strip.text = element_text(face = "bold", size = 25))+           #adjust headings
  scale_fill_manual(values = sample_colors)+                            #set fill colors
  theme(plot.title=element_text(size = 25, face = "bold", hjust = 0.5))
ggsave("figures/alpha_LDL.png",alpha_LDL)

### Generate and print statistical scores
meta = sample_data(phyloseq_object_final_highcount)
alphadiv <- estimate_richness(phyloseq_object_final_highcount, measures = c("Observed", "Shannon", "Chao1", "ACE", "Simpson", "InvSimpson", "Fisher")) %>% rownames_to_column(var = "legend") 
colnames(alphadiv)[1] = "uniqueID"
meta_alpha = inner_join(meta, alphadiv, by = "uniqueID")
meta_alpha = do.call(cbind.data.frame, meta_alpha)

res.aov2 <- aov(Shannon ~ legend, data = meta_alpha)
summary(res.aov2)
TukeyHSD(res.aov2, which = "legend")
res.aov3 <- aov(Observed ~ legend, data = meta_alpha)
summary(res.aov3)
TukeyHSD(res.aov3, which = "legend")
res.aov4 <- aov(Chao1 ~ legend, data = meta_alpha)
summary(res.aov4)
TukeyHSD(res.aov4, which = "legend")
res.aov5 <- aov(ACE ~ legend, data = meta_alpha)
summary(res.aov5)
TukeyHSD(res.aov5, which = "legend")
res.aov6 <- aov(Simpson ~ legend, data = meta_alpha)
summary(res.aov6)
TukeyHSD(res.aov6, which = "legend")
res.aov7 <- aov(InvSimpson ~ legend, data = meta_alpha)
summary(res.aov7)
TukeyHSD(res.aov7, which = "legend")
res.aov8 <- aov(Fisher ~ legend, data = meta_alpha)
summary(res.aov8)
TukeyHSD(res.aov8, which = "legend")
```

### Alpha diversity plots and stats for Fiber (Supplemental Figure 2)
```{r, message=FALSE}
### Generate alpha-diversity plot for Fiber categories
alpha_fiber <- phyloseq_object_final_highcount %>%                                                              
  plot_richness(x = "legend2",measures = diversity_measures) +                              
  geom_boxplot(aes(fill = legend2), show.legend = FALSE)+             #make violin plot, set fill aes to sampletype
  theme_linedraw()+                                                     #change theme to classic
  xlab(NULL)+                                                           #no label on x-axis
  theme(axis.text.y.left = element_text(size = 20),                     #adjust y-axis text
        axis.text.x = element_text(size = 20, hjust = 0.5, vjust = 1, angle = -90),           #adjust x-axis label position
        axis.title.y = element_text(size = 20))+                        #adjust y-axis title
  theme(strip.text = element_text(face = "bold", size = 25))+           #adjust headings
  scale_fill_manual(values = sample_colors)+                            #set fill colors
  theme(plot.title=element_text(size = 25, face = "bold", hjust = 0.5))
ggsave("figures/alpha_fiber.png",alpha_fiber)

### Generate and print statistical scores
meta = sample_data(phyloseq_object_final_highcount)
alphadiv <- estimate_richness(phyloseq_object_final_highcount, measures = c("Observed", "Shannon", "Chao1", "ACE", "Simpson", "InvSimpson", "Fisher")) %>% rownames_to_column(var = "legend2") 
colnames(alphadiv)[1] = "uniqueID"
meta_alpha = inner_join(meta, alphadiv, by = "uniqueID")
meta_alpha = do.call(cbind.data.frame, meta_alpha)

res.aov20 <- aov(Shannon ~ legend2, data = meta_alpha)
summary(res.aov20)
TukeyHSD(res.aov20, which = "legend2")
res.aov21 <- aov(Observed ~ legend2, data = meta_alpha)
summary(res.aov21)
TukeyHSD(res.aov21, which = "legend2")
res.aov22 <- aov(Chao1 ~ legend2, data = meta_alpha)
summary(res.aov22)
TukeyHSD(res.aov22, which = "legend2")
res.aov23 <- aov(ACE ~ legend2, data = meta_alpha)
summary(res.aov23)
TukeyHSD(res.aov23, which = "legend2")
res.aov24 <- aov(Simpson ~ legend2, data = meta_alpha)
summary(res.aov24)
TukeyHSD(res.aov24, which = "legend2")
res.aov25 <- aov(Simpson ~ legend2, data = meta_alpha)
summary(res.aov25)
TukeyHSD(res.aov25, which = "legend2")

head(alphadiv)
```

#BETA DIVERSITY ANALYSIS  ##############################
### Processing in different beta analysis methods using the phyloseq package 
```{r, message=FALSE}
load("phyloseq_object_final_highcount.RData")
samp_dat_wdiv <- data.frame(sample_data(phyloseq_object_final_highcount),
                            estimate_richness(phyloseq_object_final_highcount))

### PERMANOVA (Permutational ANOVA) analysis
# Use phyloseq to calculate distance matrix
dm_weighted_unifrac <- UniFrac(phyloseq_object_final_highcount, weighted=TRUE)#weighted unifrac
dm_unweighted_unifrac <- UniFrac(phyloseq_object_final_highcount, weighted=FALSE)#unweighted unifrac
dm_bray <- vegdist(t(otu_table(phyloseq_object_final_highcount)), method="bray")#bray-curtis
dm_jaccard <- vegdist(t(otu_table(phyloseq_object_final_highcount)), method="jaccard")#jaccard
# permanova: smoker-fibre_category
adonis2(dm_weighted_unifrac ~ `smoker`*fibre_category, data=samp_dat_wdiv)#weighted unifrac
adonis2(dm_unweighted_unifrac ~ `smoker`*fibre_category, data=samp_dat_wdiv)#unweighted unifrac
adonis2(dm_bray ~ `smoker`*fibre_category, data=samp_dat_wdiv)#bray-curtis
adonis2(dm_jaccard ~ `smoker`*fibre_category, data=samp_dat_wdiv)#jaccard
# permanova: smoker-LDL_category
adonis2(dm_weighted_unifrac ~ `smoker`*LDL_category, data=samp_dat_wdiv)#weighted unifrac
adonis2(dm_unweighted_unifrac ~ `smoker`*LDL_category, data=samp_dat_wdiv)#unweighted unifrac
adonis2(dm_bray ~ `smoker`*LDL_category, data=samp_dat_wdiv)#bray-curtis
adonis2(dm_jaccard ~ `smoker`*LDL_category, data=samp_dat_wdiv)#jaccard
```
## Creating PCA plots for effect of LDL in smokers, nonsmokers (Manuscript figure 2)
```{r, message=FALSE}
# Beta Diversity figure panel will include Weighted Unifrac PCoA Plot of the smoker and LDL categories, 
# divided into 2 plots, one with smokers only and the other with nonsmokers only
## Filter smokers only and nonsmokers only ##
phyloseq_object_smoking <- subset_samples(phyloseq_object_final_highcount, smoker == "Yes")
phyloseq_object_nonsmoking <- subset_samples(phyloseq_object_final_highcount, smoker =="No")
## Generate PCoA plots using weighted unifrac ##
ord.weighted_unifrac_smoking <- ordinate(phyloseq_object_smoking, method="PCoA", distance="unifrac", weighted=TRUE)
gg_pcoa_wunifrac_smoking_LDL <- plot_ordination(phyloseq_object_smoking, ord.weighted_unifrac_smoking, color = "LDL_category") +
  labs(col="LDL category") + stat_ellipse(type = "norm")#smoking-LDL
ord.weighted_unifrac_nonsmoking <- ordinate(phyloseq_object_nonsmoking, method="PCoA", distance="unifrac", weighted=TRUE)
gg_pcoa_wunifrac_nonsmoking_LDL <- plot_ordination(phyloseq_object_nonsmoking, ord.weighted_unifrac_nonsmoking, color = "LDL_category") +
  labs(col="LDL category") + stat_ellipse(type = "norm")#nonsmoking-LDL

ggsave(filename="figures/gg_pcoa_wunifrac_smoking_LDL.png",plot=gg_pcoa_wunifrac_smoking_LDL)
ggsave(filename="figures/gg_pcoa_wunifrac_nonsmoking_LDL.png",plot=gg_pcoa_wunifrac_nonsmoking_LDL)
```
## Creating PCA plots between all smoking, LDL, fiber categories (Supplemental Figure 2)
```{r, message=FALSE}
#Making a new column in the metadata to combined fibre and smoking
sample_data(phyloseq_object_final_highcount)[[40]] <-  paste(sample_data(phyloseq_object_final_highcount)[[32]],sample_data(phyloseq_object_final_highcount)[[38]])
#Making a new column in the metadata to combined LDL and smoking
sample_data(phyloseq_object_final_highcount)[[41]] <-  paste(sample_data(phyloseq_object_final_highcount)[[32]],sample_data(phyloseq_object_final_highcount)[[37]])

### PCoA plot ###
# weighted unifrac
ord.weighted_unifrac <- ordinate(phyloseq_object_final_highcount, method="PCoA", distance="unifrac", weighted=TRUE)
gg_pcoa_wunifrac_fibre <- plot_ordination(phyloseq_object_final_highcount, ord.weighted_unifrac, color = "V40") +
  labs(col="smoker, fibre category") + stat_ellipse(type = "norm")#smoker-fibre
gg_pcoa_wunifrac_LDL <- plot_ordination(phyloseq_object_final_highcount, ord.weighted_unifrac, color = "V41") +
  labs(col="smoker, LDL category") + stat_ellipse(type = "norm")#smoker-LDL
# unweighted unifrac
ord.unweighted_unifrac <- ordinate(phyloseq_object_final_highcount, method="PCoA", distance="unifrac", weighted=FALSE)
gg_pcoa_unwunifrac_fibre <- plot_ordination(phyloseq_object_final_highcount, ord.unweighted_unifrac, color = "V40") +
  labs(col="smoker, fibre category") + stat_ellipse(type = "norm")#smoker-fibre
gg_pcoa_unwunifrac_LDL <- plot_ordination(phyloseq_object_final_highcount, ord.unweighted_unifrac, color = "V41") +
  labs(col="smoker, LDL category") + stat_ellipse(type = "norm")#smoker-LDL
# bray-curtis
ord.bray <- ordinate(phyloseq_object_final_highcount, method="PCoA", distance="bray")
gg_pcoa_bray_fibre <- plot_ordination(phyloseq_object_final_highcount, ord.bray, color = "V40") +
  labs(col="smoker, fibre category") + stat_ellipse(type = "norm")#smoker-fibre
gg_pcoa_bray_LDL <- plot_ordination(phyloseq_object_final_highcount, ord.bray, color = "V41") +
  labs(col="smoker, LDL category") + stat_ellipse(type = "norm")#smoker-LDL

ggsave(filename="figures/gg_pcoa_wunifrac_fibre.png" , plot=gg_pcoa_wunifrac_fibre)
ggsave(filename="figures/gg_pcoa_wunifrac_LDL.png", plot= gg_pcoa_wunifrac_LDL)
ggsave(filename="figures/gg_pcoa_unwunifrac_fibre.png", plot=gg_pcoa_unwunifrac_fibre)
ggsave(filename="figures/gg_pcoa_unwunifrac_LDL.png", plot=gg_pcoa_unwunifrac_LDL)
ggsave(filename="figures/gg_pcoa_bray_fibre.png", plot=gg_pcoa_bray_fibre)
ggsave(filename="figures/gg_pcoa_bray_LDL.png", plot=gg_pcoa_bray_LDL)
```


# DESEQ2 ANALYSIS ##############################
## Processing phyloseq object into DESeq2 (smoking)
```{r, message=FALSE}
### Tidy up phyloseq object, relevel to set low LDL as baseline
phyloseq_object_final_plus1 <- transform_sample_counts(phylo_smoking, function(x) x+1)
phyloseq_object_final_deseq <- phyloseq_to_deseq2(phyloseq_object_final_plus1, ~ relevel(LDL_category, "low")) 
phyloseq_final <- DESeq(phyloseq_object_final_deseq)
### Ensure results is the healthy group
res <- results(phyloseq_final, tidy=TRUE)
colnames(res)[1] = "ASV"
res_with_taxa = inner_join(taxa_info,res, by = "ASV" )

### Get significant hits
res_sig = res_with_taxa %>% filter( padj<0.01 & abs(log2FoldChange)>2)

### count and report ASVs significantly up/down reg in smokers 
upregulated_count <- sum(res_sig$log2FoldChange > 0)
downregulated_count <- sum(res_sig$log2FoldChange < 0)
cat("Number of upregulated smoking ASVs:", upregulated_count, "\n")
cat("Number of downregulated smoking ASVs:", downregulated_count, "\n")

### Order results by log2FC
res_sig <- res_sig[order(res_sig$log2FoldChange),]
### Function to combine p values
combine_pvalues = function(p){return(1-pchisq(-2*sum(log(p),na.rm=T),2*sum(!is.na(p))))}

### Translating from ASV to genus level by merging the data for rows that have the same genus
### Need to take the average for log2FC and combined the Pvalues
res_genus_combined = res_sig %>%
  group_by(Genus) %>%
  summarize(log2FoldChange_avg = mean(log2FoldChange), pvalues =  combine_pvalues(pvalue)) %>%
  na.omit()
res_genus_combined <- res_genus_combined[order(res_genus_combined$log2FoldChange_avg),]

```
## Using DESeq2 data to produce barplots + volcano plots (smoking) (Manuscript figure 3) 
```{r}
sighits_smok = ggplot(data = res_genus_combined, aes(x= log2FoldChange_avg,y=reorder(Genus, -(as.numeric(log2FoldChange_avg))), fill = pvalues))+
  geom_bar(stat = "identity") +
  theme_bw()+
  scale_fill_gradient(low = "yellow", high = "red", na.value = NA)
ggsave("figures/smoking_LDL_phyloseq_DeSeq.png", sighits_smok)

vol_plot_smok <- res %>%
  mutate(significant = padj<0.01 & abs(log2FoldChange)>2) %>% #new column in results table 
  ggplot() +
  geom_point(aes(x=log2FoldChange, y=-log10(padj), col=significant)) + 
  labs(x = "Log2 Fold Change", y = "-Log10 (padj)")
ggsave(filename="figures/vol_plot_smoking_LDL.png",vol_plot_smok)
```
## Processing phyloseq object into DESeq2 (NONsmoking)
```{r, message=FALSE}
### Tidy up phyloseq object, relevel to set low LDL as baseline
phyloseq_object_final_plus1 <- transform_sample_counts(phylo_nonsmoking, function(x) x+1)
phyloseq_object_final_deseq <- phyloseq_to_deseq2(phyloseq_object_final_plus1, ~ relevel(LDL_category, "low")) 
phyloseq_final <- DESeq(phyloseq_object_final_deseq)
### Ensure results is the healthy group
res_nonsmoke <- results(phyloseq_final, tidy=TRUE)
colnames(res_nonsmoke)[1] = "ASV"
res_nonsmoke_taxa = inner_join(taxa_info,res_nonsmoke, by = "ASV" )

### Get significant hits
res_nonsmoke_sig = res_nonsmoke_taxa %>% filter( padj<0.01 & abs(log2FoldChange)>2)

### count and report ASVs significantly up/down reg in smokers 
upregulated_count <- sum(res_nonsmoke_sig$log2FoldChange > 0)
downregulated_count <- sum(res_nonsmoke_sig$log2FoldChange < 0)
cat("Number of upregulated nonsmoking ASVs:", upregulated_count, "\n")
cat("Number of downregulated nonsmoking ASVs:", downregulated_count, "\n")

### Order results by log2FC
res_nonsmoke_sig <- res_nonsmoke_sig[order(res_nonsmoke_sig$log2FoldChange),]

### Translating from ASV to genus level by merging the data for rows that have the same genus
### Need to take the average for log2FC and combined the Pvalues
res_nonsmoke_genus_combined = res_nonsmoke_sig %>%
  group_by(Genus) %>%
  summarize(log2FoldChange_avg = mean(log2FoldChange), pvalues =  combine_pvalues(pvalue)) %>%
  na.omit()
res_nonsmoke_genus_combined <- res_nonsmoke_genus_combined[order(res_nonsmoke_genus_combined$log2FoldChange_avg),]
```
## Using DESeq2 data to produce barplots + volcano plots (NONsmoking) (Supplemental Figure 3) 
```{r}
sighits_nonsmok = ggplot(data = res_nonsmoke_genus_combined, aes(x= log2FoldChange_avg,y=reorder(Genus, -(as.numeric(log2FoldChange_avg))), fill = pvalues))+
  geom_bar(stat = "identity") +
  theme_bw()+
  scale_fill_gradient(low = "yellow", high = "red", na.value = NA)
ggsave(filename="figures/nonsmoking_LDL_phyloseq_DeSeq.png", sighits_nonsmok)

vol_plot_nonsmok <- res_nonsmoke %>%
  mutate(significant = padj<0.01 & abs(log2FoldChange)>2) %>% #new column in results table 
  ggplot() +
  geom_point(aes(x=log2FoldChange, y=-log10(padj), col=significant)) + 
  labs(x = "Log2 Fold Change", y = "-Log10 (padj)")
ggsave(filename="figures/vol_plot_nonsmoking_LDL.png",vol_plot_nonsmok)
```

# PICRUST ANALYSIS   ##############################
### Setup by loading in data and processing for functional analyses
```{r, message=FALSE}
### PICRUSt2 is pickier about the formatting of the data, so first run this block to ensure everything is as needed.
###Load a custom function by Chris to provide all DESeq2 results
source("DESeq2_function.R")
###Removing all columns that include NA's for the pathways_daa() function
metadata = metadata[ , colSums(is.na(metadata)) == 0]
###Filter your metadata as needed to look at specific comparisons
metadata_smoking = metadata %>%filter(smoker == "Yes")
metadata_nonsmoking = metadata %>%filter(smoker == "No")

###Filtering the abundance table to only include samples that are in the filtered metadata
sample_names_smoking = metadata_smoking$`#SampleID`
sample_names_smoking = append(sample_names_smoking, "pathway")
abundance_data_smoking = abundance_data[, colnames(abundance_data) %in% sample_names_smoking] 

sample_names_nonsmoking = metadata_nonsmoking$`#SampleID`
sample_names_nonsmoking = append(sample_names_nonsmoking, "pathway")
abundance_data_nonsmoking = abundance_data[, colnames(abundance_data) %in% sample_names_nonsmoking] 

###Removing individuals with no data that caused a problem for pathways_daa()
abundance_data_smoking =  abundance_data_smoking[, colSums(abundance_data_smoking != 0) > 0]
abundance_data_nonsmoking =  abundance_data_nonsmoking[, colSums(abundance_data_nonsmoking != 0) > 0]

###Filtering out columns that represent a total abundance < 10000
###Filtering out rows (pathways) that have a total count < 100
abundance_data_smoking = abundance_data_smoking[,colSums(abundance_data_smoking[,-1]) > 10000]
abundance_data_smoking = abundance_data_smoking[rowSums(abundance_data_smoking[,-1]) > 100,]
abundance_data_nonsmoking = abundance_data_nonsmoking[,colSums(abundance_data_nonsmoking[,-1]) > 10000]
abundance_data_nonsmoking = abundance_data_nonsmoking[rowSums(abundance_data_nonsmoking[,-1]) > 100,]

###Ensuring the rownames for the abundance_data_filtered is empty. This is required for their functions to run.
rownames(abundance_data_smoking) = NULL
rownames(abundance_data_nonsmoking) = NULL

###Verify samples in metadata match samples in abundance_data
abun_samples_smoking = rownames(t(abundance_data_smoking[,-1])) #Get list of sample names in  newly filtered abundance data
metadata_smoking = metadata_smoking[metadata_smoking$`#SampleID` %in% abun_samples_smoking,] #Ensure filtered metadata only includes these samples
abun_samples_nonsmoking = rownames(t(abundance_data_nonsmoking[,-1]))
metadata_nonsmoking = metadata_nonsmoking[metadata_nonsmoking$`#SampleID` %in% abun_samples_nonsmoking,]

### Filter data to remove 0 standard deviation
row_sds_nonsmok <- abundance_data_nonsmoking %>% column_to_rownames("pathway") %>% apply(1, sd)
filtered_LDL_nonsmok <- abundance_data_nonsmoking[row_sds_nonsmok != 0, ]
row.names(filtered_LDL_nonsmok) <- NULL
row_sds_smok <- abundance_data_smoking %>% column_to_rownames("pathway") %>% apply(1, sd)
filtered_LDL_smok <- abundance_data_smoking[row_sds_smok != 0, ]
row.names(filtered_LDL_smok) <- NULL

# Perform pathway DAA using DESeq2 method
metacyc_daa_results_df_smoking <- pathway_daa(abundance = abundance_data_smoking %>% column_to_rownames("pathway"), metadata = metadata_smoking, group = "LDL_category", daa_method = "DESeq2")
# Annotate MetaCyc pathway results without KO to KEGG conversion. This line provides a useful description for your pathway accessions. 
metacyc_daa_annotated_results_df_smoking <- pathway_annotation(pathway = "MetaCyc", daa_results_df = metacyc_daa_results_df_smoking, ko_to_kegg = FALSE)

metacyc_daa_results_df_nonsmoking <- pathway_daa(abundance = abundance_data_nonsmoking %>% column_to_rownames("pathway"), metadata = metadata_nonsmoking, group = "LDL_category", daa_method = "DESeq2")
metacyc_daa_annotated_results_df_nonsmoking <- pathway_annotation(pathway = "MetaCyc", daa_results_df = metacyc_daa_results_df_nonsmoking, ko_to_kegg = FALSE)

#Generating the full DESeq2 results dataframe
#Here is where we can see HOW up or downregulated a certain pathway is.
#The reference group is selected alphabetically. In this case, Dogs are the reference and anything with a positive log2foldchange means it is higher in humans.
smoking_res = DEseq2_function(abundance_data_smoking, metadata_smoking)
smoking_res$feature = rownames(smoking_res)
nonsmoking_res = DEseq2_function(abundance_data_nonsmoking, metadata_nonsmoking)
nonsmoking_res$feature = rownames(nonsmoking_res)

#Joining description
smoking_res_desc = inner_join(smoking_res, metacyc_daa_annotated_results_df_smoking, by = "feature")
smoking_res_desc = smoking_res_desc[, -c(8:13)]
nonsmoking_res_desc = inner_join(nonsmoking_res, metacyc_daa_annotated_results_df_nonsmoking, by = "feature")
nonsmoking_res_desc = nonsmoking_res_desc[, -c(8:13)]

#Filter by p-value <0.05
smoking_res_desc_sig = smoking_res_desc %>% filter(pvalue <0.05)
nonsmoking_res_desc_sig = nonsmoking_res_desc %>% filter(pvalue <0.05)
```
### Graph PICRUSt results in barplots, filtered for >1 log2 fold chance (Manuscript Figure 4)
```{r}
smoking_res_desc_sig <- smoking_res_desc_sig[order(smoking_res_desc_sig$log2FoldChange),]
nonsmoking_res_desc_sig <- nonsmoking_res_desc_sig[order(nonsmoking_res_desc_sig$log2FoldChange),]

###Filtering for >1 fold change
smoking_res_above1 <- smoking_res_desc_sig %>%
  filter(log2FoldChange > 1 | log2FoldChange < -1) %>% arrange(log2FoldChange)
nonsmoking_res_above1 <- nonsmoking_res_desc_sig %>%
  filter(log2FoldChange > 1 | log2FoldChange < -1) %>% arrange(log2FoldChange)

### Barplot for smoking results
picrust_barplot_smoking <- ggplot(data = smoking_res_above1, aes(x= log2FoldChange,y=reorder(description, -(as.numeric(log2FoldChange))), 
                                                                 fill = pvalue))+ theme_bw()+
  geom_bar(stat = "identity") +
  scale_fill_gradient(low = "yellow", high = "red", na.value = NA)+
  xlim(-0.5,2.5)+
  labs(x = "log2 Fold Change", y = "Bacterial Pathway")+
  ggtitle("LDL-Associated Pathway Changes in Smokers")

### Barplot for nonsmoking results
picrust_barplot_nonsmoking <- ggplot(data = nonsmoking_res_above1, aes(x= log2FoldChange,y=reorder(description, -(as.numeric(log2FoldChange))), 
                                         fill = pvalue))+ theme_bw()+
  geom_bar(stat = "identity") +
  scale_fill_gradient(low = "yellow", high = "red", na.value = NA)+
  xlim(-0.5,2.5)+
  labs(x = "log2 Fold Change", y = "Bacterial Pathway")+
  ggtitle("LDL-Associated Pathway Changes in Non-Smokers")

### Barplot for smoking + nonsmoking results combined
smoking_res_above1$smoking="S"
nonsmoking_res_above1$smoking="NS"
figure4_res_above1 <- rbind(smoking_res_above1, nonsmoking_res_above1)
figure4_res_above1$smoking = factor(figure4_res_above1$smoking, levels=c("S","NS"))

picrust_barplot_all <- ggplot(data = figure4_res_above1, aes(x= log2FoldChange,y=reorder(description, -(as.numeric(log2FoldChange))), 
                                      fill = pvalue))+ theme_bw()+
  geom_bar(stat = "identity") +
  scale_fill_gradient(low = "yellow", high = "red", na.value = NA)+
  xlim(-0.5,2.5)+
  labs(x = "Log2 Fold Change", y = "Bacterial Pathway")+
  facet_grid(rows = vars(smoking),scales ="free_y", space = "free_y")+
  theme(strip.text = element_text(
    size = 10, color = "black"), 
    axis.text.y = element_text(size = 18, face = "bold"), 
    legend.text = element_text(size = 12, face = "bold"))

ggsave(filename="figures/picrust_barplot_smoking.png",picrust_barplot_smoking,height=10,width=15)
ggsave(filename="figures/picrust_barplot_nonsmoking.png",picrust_barplot_nonsmoking,height=6,width=12)
ggsave(filename="figures/picrust_barplot_all.png",picrust_barplot_all,height=10,width=15)
```

### Using ggpicrust to create PCA plots (Supplemental Figure 4)
```{r, message=FALSE}
PCA_smoking_all <- pathway_pca(abundance = abundance_data %>% column_to_rownames("pathway"), 
                         metadata = metadata, group = "smoker")+
                        ggtitle("PCA plot comparing smoking in all participants")
PCA_LDL_all <- pathway_pca(abundance = abundance_data %>% column_to_rownames("pathway"), 
                           metadata = metadata, group = "LDL_category")+
                        ggtitle("PCA plot comparing LDL in all participants")
PCA_nonsmok_LDL <- pathway_pca(abundance = filtered_LDL_nonsmok %>% column_to_rownames("pathway"), 
                               metadata = metadata_nonsmoking, group = "LDL_category")+
                                ggtitle("            Non-smokers")
PCA_smok_LDL <- pathway_pca(abundance = filtered_LDL_smok %>% column_to_rownames("pathway"), 
                            metadata = metadata_smoking, group = "LDL_category")+
                                ggtitle("            Smokers")

ggsave(filename="figures/picrust_PCA_smoking_all.png",PCA_smoking_all)
ggsave(filename="figures/picrust_PCA_LDL_all.png",PCA_LDL_all)
ggsave(filename="figures/picrust_PCA_nonsmok_LD.png",PCA_nonsmok_LDL)
ggsave(filename="figures/picrust_PCA_smok_LDL.png",PCA_smok_LDL)

```